sockeye/layers.py

```diff
@@ -46,11 +48,1542 @@ def activation(data: mx.sym.Symbol, act_type: str) -> mx.sym.Symbol:
+                 max_length: Optional[int] = None) -> mx.sym.Symbol:
+        assert max_length is not None, "max_length needed for positional embeddings."
+        # (1, source_seq_len)
+        positions = mx.sym.expand_dims(data=mx.sym.arange(start=0, stop=max_length, step=1), axis=0)
+
+        # (1, source_seq_len, num_embed)
+        pos_embedding = mx.sym.Embedding(data=positions,
+                                         input_dim=self.max_seq_len,
+                                         weight=self.embed_weight,
+                                         output_dim=self.num_embed,
+                                         name=self.prefix + "pos_embed")
+        data = mx.sym.broadcast_add(data, pos_embedding, name="%s_add" % self.prefix)
+        if self.dropout > 0.0:
+            data = mx.sym.Dropout(data, p=self.dropout)
+        return data
+
+    def step(self, step: int, data: mx.sym.Symbol) -> mx.sym.Symbol:
+        position = step - 1
+        position = position * mx.sym.reshape(
+            mx.sym.slice_axis(mx.sym.reshape(mx.sym.ones_like(data), shape=(0, -1)), axis=1, begin=0, end=1),
+            shape=(-1))
+        pos_embedding = mx.sym.Embedding(data=position,
+                                         input_dim=self.max_seq_len,
+                                         weight=self.embed_weight,
+                                         output_dim=self.num_embed,
+                                         name=self.prefix + "pos_embed")
+        pos_embedding = mx.sym.expand_dims(pos_embedding, axis=1)
+        return mx.sym.broadcast_add(data, pos_embedding, name="%s_add" % self.prefix)
+
+    def get_num_hidden(self) -> int:
+        return self.num_embed
+
+    def get_max_seq_len(self):
+        return self.max_seq_len
+
+
+class LearnedPositionalEmbeddingsLayerConfig(LayerConfig):
+
+    def __init__(self,
+                 num_embed: int,
+                 dropout: float,
+                 max_seq_len: int):
+        super().__init__()
+        self.num_embed = num_embed
+        self.dropout = dropout
+        self.max_seq_len = max_seq_len
+
+    def create_encoder_layer(self, input_num_hidden: int, prefix: str) -> EncoderLayer:
+        return self.create_layer(prefix)
+
+    def create_decoder_layer(self, input_num_hidden: int, prefix: str) -> DecoderLayer:
+        return self.create_layer(prefix)
+
+    def create_layer(self, prefix: str) -> StatelessBlockLayer:
+        return StatelessBlockLayer(LearnedAdditivePositionalEmbeddings(num_embed=self.num_embed,
+                                                                       dropout=self.dropout,
+                                                                       max_seq_len=self.max_seq_len,
+                                                                       prefix=prefix + "pos_embed_"))
+
+
+# TODO: share the code with the encoder!
+class AdditiveSinCosPositionalEmbeddings(StatelessBlock):
```

Comment:
can we try to use these blocks in the existing models as well to share code? I know that TransformerDecoder is not a CustomSeqDecoder but maybe we can avoid duplicating implementations.

Reply:
+1. I was planning on refactoring the code this way (see the `TODO: share the code with the encoder!`).